MONIT-5441: Added code to handle source tags and descriptions.

spramanik4667 opened the pull request:
Added Deocoder and Handler code for sourceTags and sourceDescription metadata.

panghy commented:
There are conflicting files, please rebase this PR

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+  };
+
+  private final List<FormatterElement> elements;
+  private final ThreadLocal<DSWrapperLexer> dsWrapperLexerThreadLocal =
```
can be private static final

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+        }
+      };
+
+  private SourceTagIngesterFormatter(List<FormatterElement> elements) {this.elements = elements;}
```
We don't usually collapse bodies (the IDE can do it)

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+
+  private SourceTagIngesterFormatter(List<FormatterElement> elements) {this.elements = elements;}
+
+  public static SourceTagIngesterFormatBuilder newBuilder() {
```
javadoc for public methods

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+    return new SourceTagIngesterFormatBuilder();
+  }
+
+  public ReportSourceTag drive(String input, String defaultHostName, String customerId,
```
javadoc for public methods

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+    return sourceTag;
+  }
+
+  public static class SourceTagIngesterFormatBuilder {
```
javadoc for public class

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+  public static class SourceTagIngesterFormatBuilder {
+    private final List<FormatterElement> elements = Lists.newArrayList();
+
+    public SourceTagIngesterFormatBuilder appendCaseSensitiveLiterals(String[] literals) {
```
ditto and for the rest of the class

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+ *
+ * @author Suranjan Pramanik (suranjan@wavefront.com).
+ */
+public class SourceTagIngesterFormatter {
```
Do you need to create another ingester formmater for this? Can you not reuse the same formatter class instead of copying a lot of code over?

panghy replied:
If you want a different output class, perhaps use an abstract class the genericize the drive method

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,324 @@
+  public static final String DESCRIPTION = "description";
+  public static final String ACTION = "action";
+  public static final String ACTION_SAVE = "save";
+  public static final String  ACTION_DELETE ="delete";
```
white space

panghy commented on proxy/src/main/java/com/wavefront/agent/PostSourceTagTimedTask.java:
```diff
@@ -0,0 +1,246 @@
+          drainBuffersToQueue();
+        }
+      }
+    }catch (Throwable t) {
```
space before catch (please use auto-format)

panghy commented on proxy/src/main/java/com/wavefront/agent/PushAgent.java:
```diff
@@ -75,6 +75,15 @@ protected void startListeners() {
       }
     }
+    // ------ Start the listener for source metadata, such as sourceTags and description ----------
+    if(metadataListenerPorts != null) {
```
space before (

panghy commented on proxy/src/main/java/com/wavefront/agent/SourceTagHandlerImpl.java:
```diff
@@ -0,0 +1,132 @@
+  /**
+   * This method will read the input over the socket.
+   *
+   * @param ctx
```
document

panghy commented on proxy/src/main/java/com/wavefront/agent/SourceTagHandlerImpl.java:
```diff
@@ -0,0 +1,132 @@
+  private static final SourceTagDecoder sourceTagDecoder = new SourceTagDecoder();
+
+  public SourceTagHandlerImpl(PostSourceTagTimedTask[] tasks) {
+    sendDataTasks = tasks;
```
this.sendDataTasks

panghy commented on proxy/src/main/java/com/wavefront/agent/SourceTagHandlerImpl.java:
```diff
@@ -0,0 +1,132 @@
+    try {
+      sourceTagDecoder.decodeSourceTagLine(sourceTagLine, sourceTags);
+    } catch (Exception ex) {
+      // TODO: Handle the exception
```
at least log?

panghy commented on proxy/src/main/java/com/wavefront/agent/SourceTagHandlerImpl.java:
```diff
@@ -0,0 +1,132 @@
+        throw new IllegalArgumentException(errorMessage);
+      }
+      // TODO: validate the sourceTag and call server API: Currently log it
+      logger.info(String.format("Message Type = %s", reportSourceTag.getSourceTagLiteral()));
```
not sure we should log all the time at info

panghy commented on proxy/src/main/java/com/wavefront/agent/SourceTagHandlerImpl.java:
```diff
@@ -0,0 +1,132 @@
+  public void reportSourceTags(List<ReportSourceTag> sourceTags) {
+    for (ReportSourceTag reportSourceTag : sourceTags) {
+      if (!annotationKeysAreValid(reportSourceTag)) {
+        String errorMessage = "WF-401 : SourceTag annotation key has illegal characters.";
```
Did you check that 401 is unused?

panghy commented on proxy/src/main/java/com/wavefront/agent/SourceTagHandlerImpl.java:
```diff
@@ -0,0 +1,132 @@
+    long min = Long.MAX_VALUE;
+    PostSourceTagTimedTask randomTask = null;
+    PostSourceTagTimedTask firstChoice = null;
+    for (int i=0; i < this.sendDataTasks.length; i++) {
```
spacing is odd (use auto-format)

spramanik4667 commented:
I have updated the PR. I have added an AbstractIngesterFormatter as base class and IngesterFormatter and SourceTagIngesterFormatter as sub classes. Also I have created an AbstractWrapper class which provides a wrapper around ReportPoint and ReportSourceTag.

panghy commented on java-lib/src/main/java/com/wavefront/ingester/HistogramDecoder.java:
```diff
@@ -16,7 +16,7 @@
 public class HistogramDecoder implements Decoder<String> {
   private static final Logger logger = Logger.getLogger(HistogramDecoder.class.getCanonicalName());
-  private static final IngesterFormatter FORMAT = IngesterFormatter.newBuilder()
+  private static final IngesterFormatter FORMAT = (IngesterFormatter) IngesterFormatter.newBuilder()
```
Please use generics for this if possible (make AbstractIngesterFormatter take a type parameter)

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagDecoder.java:
```diff
@@ -0,0 +1,36 @@
+  private static final String SOURCE_DESCRIPTION = "@SourceDescription";
+
+  private static final SourceTagIngesterFormatter FORMAT = (SourceTagIngesterFormatter)
+      SourceTagIngesterFormatter
```
ditto re: generics, also we wouldn't typically line break with just a class without a method call (remove new line on line 21)

panghy commented on proxy/src/main/java/com/wavefront/agent/QueuedAgentService.java:
```diff
@@ -385,6 +312,163 @@ public Long value() {
+          @Override
+          public ResubmissionTask from(byte[] bytes) throws IOException {
+            try {
+              if (bytes.length > 2 && bytes[0] == (byte) 0x1f && bytes[1] == (byte) 0x8b) {
```
you can probably remove logic that does this (for that matter, we should remove this logic completely) since this is for backwards compatibility and we shouldn't need any

panghy commented on proxy/src/test/java/com/wavefront/agent/SourceTagHandlerTest.java:
```diff
@@ -0,0 +1,101 @@
+ */
+public class SourceTagHandlerTest {
+
+  private static final Logger logger = Logger.getLogger(SourceTagHandlerTest.class
```
what is the logger for?

spramanik4667 commented:
I have removed the 'GZip' logic detection and added generics to the AbstractIngesterFormatter class.

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,94 @@
+   */
+  @Override
+  public T drive(String input, String defaultHostName, String customerId,
+                               List<String> customerSourceTags) {
```
align params

panghy commented on java-lib/src/main/java/com/wavefront/ingester/SourceTagIngesterFormatter.java:
```diff
@@ -0,0 +1,94 @@
+   * @param defaultHostName
+   * @param customerId
+   * @param customerSourceTags
+   * @return
```
remove or comment (check for empty @return in this file)

spramanik4667 commented:
Added the javadoc comments.

panghy commented:
Sorry for the delay, please resolve conflicting files and we'll proceed

spramanik4667 commented:
Resolved the conflicts.

panghy commented:
I am going to merge this in but on second thought, we should be able to do this on a single port rather than on two (same is true for aggregated histograms).

Basically if it starts with "@" it's a metadata operation and "!" is a histogram.

That's outside the scope of this ticket but I'll file an issue and assign it to you (should you find yourself wanting a challenge).
